### ISSUE TITLE
tailscaled dns: error: upstream nameservers not set 

### ISSUE THREAD

**Issue by christopher-cavnue** [tailscale/tailscale#3848]
### What is the issue?

many, many of these messages.

Jan 30 16:53:56 field-be-002-nx2 tailscaled dns: error: upstream nameservers not set
Jan 30 16:53:56 field-be-002-nx2 tailscaled dns: error: upstream nameservers not set
Jan 30 16:53:56 field-be-002-nx2 tailscaled [RATELIMIT] format("dns: error: %v")

cavnue@field-be-002-nx2:~$ sudo tailscale bugreport
BUG-7346ad9239ae6805630a693592f8fed31b0357f26255f1263fff2fe217a386d9-20220131005448Z-f082a39f8b09848d
cavnue@field-be-002-nx2:~$ 

### Steps to reproduce

not sure

### Are there any recent changes that introduced the issue?

no.

### OS

Linux

### OS version

18.04.5 LTS on aarch64

### Tailscale version

1.20.1

### Bug report

BUG-7346ad9239ae6805630a693592f8fed31b0357f26255f1263fff2fe217a386d9-20220131005448Z-f082a39f8b09848d

**Comment by cepera-ang** [comment]
I have seen this error too. After checking "Override local DNS" in MagicDNS settings it went away (but this feels more like workaround than actual investigation and solution).

**Comment by DentonGentry** [comment]
Looking at BUG-7346ad9239ae6805630a693592f8fed31b0357f26255f1263fff2fe217a386d9-20220131005448Z-f082a39f8b09848d

The first time we see "upstream nameservers not set"

```
2022-01-25 15:15:24.532245135 +0000 UTC: wgengine: Reconfig: configuring DNS
2022-01-25 15:15:24.532466288 +0000 UTC: dns: Set: {DefaultResolvers:[] Routes:{example.com.beta.tailscale.net.:[]}+65arpa SearchDomains:[example.com.beta.tailscale.net.] Hosts:9}
...
2022-01-25 15:15:28.032957519 +0000 UTC: dns: error: upstream nameservers not set
2022-01-25 15:15:28.033503187 +0000 UTC: dns: error: upstream nameservers not set
```

Prior to that, tailscaled had just started and appears to have had no network connectivity at first:
```
2021-07-21 18:51:41.125982034 +0000 UTC: Program starting: v1.20.1-t6dbdbb41a-g41b221f94, Go 1.17.6-tse44d304e54: []string{"/usr/sbin/tailscaled", "--cleanup"}
2021-07-21 18:51:41.127837374 +0000 UTC: LogID: 7346ad9239ae6805630a693592f8fed31b0357f26255f1263fff2fe217a386d9
2021-07-21 18:51:41.127876383 +0000 UTC: logpolicy: using system state directory "/var/lib/tailscale"
2021-07-21 18:51:41.129743276 +0000 UTC: logtail: dial "log.tailscale.io:443" failed: dial tcp: lookup log.tailscale.io on 127.0.0.53:53: server misbehaving (in 2ms), trying bootstrap...
2021-07-21 18:51:41.133995497 +0000 UTC: dns: [rc=resolved nm=no ret=systemd-resolved]
2021-07-21 18:51:41.13410481 +0000 UTC: trying bootstrapDNS("derp10.tailscale.com", "137.220.36.168") for "log.tailscale.io" ...
2021-07-21 18:51:41.134332236 +0000 UTC: creating dns cleanup: route ip+net: no such network interface
2021-07-21 18:51:41.134380652 +0000 UTC: flushing log.
2021-07-21 18:51:41.134420172 +0000 UTC: logger closing down
2021-07-21 18:51:41.13458843 +0000 UTC: bootstrapDNS("derp10.tailscale.com", "137.220.36.168") for "log.tailscale.io" error: Get "https://derp10.tailscale.com/bootstrap-dns?q=log.tailscale.io": dial tcp 137.220.36.168:443: connect: network is unreachable
2021-07-21 18:51:41.134652302 +0000 UTC: trying bootstrapDNS("derp5.tailscale.com", "2001:19f0:5801:10b7:5400:2ff:feaa:284c") for "log.tailscale.io" ...
2021-07-21 18:51:41.134935952 +0000 UTC: bootstrapDNS("derp5.tailscale.com", "2001:19f0:5801:10b7:5400:2ff:feaa:284c") for "log.tailscale.io" error: Get "https://derp5.tailscale.com/bootstrap-dns?q=log.tailscale.io": dial tcp [2001:19f0:5801:10b7:5400:2ff:feaa:284c]:443: connect: network is unreachable
```

So presumably:
- it was using dns.DirectManager
- /etc/resolv.conf contained no nameservers, because there was no connectivity

**Comment by christopher-cavnue** [comment]
i'm able to access the device ok.   (it's on LTE/5G.)

BUG-7346ad9239ae6805630a693592f8fed31b0357f26255f1263fff2fe217a386d9-20220131071639Z-921f8cf69c4cc8c5



Christopher Hoover  |  V.P. Hardware Engineering

www.cavnue.com

**Comment by DentonGentry** [comment]
For those experiencing a problem with /etc/resolv.conf and detecting systemd-resolved, could you try installing from the unstable track https://pkgs.tailscale.com/unstable/ build 1.21.73 or later?

[#3905](https://github.com/tailscale/tailscale/pull/3905) reorders some of the steps in tailscaled's startup process to try to ensure that systemd-resolved is initialized and ready to respond before tailscaled checks it.

**Comment by christopher-cavnue** [comment]
we've put our field devices on --accept-dns=false.   that's  the right
thing to do for our application -- there's no point in our field devices
running magicdns.


Christopher Hoover  |  V.P. Hardware Engineerifalse

www.cavnue.com

**Comment by DentonGentry** [comment]
Fixed in 1.22 by better detecting systemd-resolved.

**Comment by cbeckr** [comment]
We're seeing this with Tailscale 1.30.1 and system-resolved - on a fresh installation of Ubuntu Server 22.04.01 running in a local network*.
Checking "Override local DNS" as described above seems to work around it successfully.

* this is relevant as there's apparently a fallback to the default EC2/GCP DNS servers in place

**Comment by cbeckr** [comment]
@DentonGentry, please let me know if we should file a new issue instead.

**Comment by DentonGentry** [comment]
Please file a new issue, I see no indication that you are hitting the same problem.
We believe we fixed this problem in 1.22.
